In [3]:
!pip install scikit-surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 18.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=2811602 sha256=491fab60f6fcd0a2adcfbfa6097c8041812cb629ed3ec5633482a22fbe247638
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [4]:
import pandas as pd
import numpy as np
from surprise import SVD, NMF, SVDpp
from surprise import Dataset
from surprise import Reader
from surprise.model_selection import cross_validate

In [5]:
ratings = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/ratings.csv')
movies = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/movies.csv')
tags = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/tags.csv')

In [6]:
merged_data = pd.merge(ratings, movies, on='movieId')
merged_data = pd.merge(merged_data, tags, on=['userId', 'movieId'])

In [7]:
reader = Reader(rating_scale=(1, 5))
data = Dataset.load_from_df(merged_data[['userId', 'movieId', 'rating']], reader)

In [8]:
svd = SVD()
svdpp = SVDpp()
nmf = NMF()

In [9]:
cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.5372  0.6044  0.5503  0.5651  0.5457  0.5605  0.0237  
MAE (testset)     0.3241  0.3645  0.3484  0.3539  0.3437  0.3469  0.0133  
Fit time          0.03    0.02    0.02    0.02    0.02    0.02    0.00    
Test time         0.00    0.00    0.00    0.00    0.00    0.00    0.00    


{'test_rmse': array([0.53717678, 0.60435828, 0.55029401, 0.56509959, 0.54574506]),
 'test_mae': array([0.3241376 , 0.36454028, 0.34841328, 0.35391831, 0.34373096]),
 'fit_time': (0.025118350982666016,
  0.023439884185791016,
  0.021068572998046875,
  0.024358510971069336,
  0.021538734436035156),
 'test_time': (0.0033326148986816406,
  0.002744913101196289,
  0.0027663707733154297,
  0.004416227340698242,
  0.002777576446533203)}

In [10]:
cross_validate(svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.5238  0.5340  0.5157  0.5381  0.5652  0.5353  0.0168  
MAE (testset)     0.2969  0.3144  0.3049  0.3108  0.3152  0.3084  0.0068  
Fit time          4.54    4.69    3.51    3.63    3.36    3.94    0.55    
Test time         0.36    0.36    0.16    0.19    0.14    0.24    0.10    


{'test_rmse': array([0.52379377, 0.53395403, 0.51568913, 0.53807846, 0.56515063]),
 'test_mae': array([0.29690076, 0.31440982, 0.30492781, 0.3107631 , 0.31519688]),
 'fit_time': (4.5413978099823,
  4.685488939285278,
  3.508847713470459,
  3.6315739154815674,
  3.3568592071533203),
 'test_time': (0.35553979873657227,
  0.36386871337890625,
  0.15796351432800293,
  0.18631219863891602,
  0.1433427333831787)}

In [11]:
cross_validate(nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.5346  0.6080  0.5751  0.6114  0.6217  0.5902  0.0319  
MAE (testset)     0.2656  0.3247  0.2909  0.3088  0.3154  0.3011  0.0209  
Fit time          0.07    0.07    0.07    0.08    0.08    0.07    0.00    
Test time         0.00    0.00    0.00    0.00    0.00    0.00    0.00    


{'test_rmse': array([0.53455745, 0.60800721, 0.57513214, 0.6113855 , 0.62174791]),
 'test_mae': array([0.26560564, 0.3246801 , 0.29086042, 0.30881146, 0.31540742]),
 'fit_time': (0.07419395446777344,
  0.07012176513671875,
  0.07219529151916504,
  0.07956123352050781,
  0.0768589973449707),
 'test_time': (0.0024271011352539062,
  0.0024034976959228516,
  0.003101825714111328,
  0.002404928207397461,
  0.0024421215057373047)}

###SVD
RMSE: 0.5372 (std: 0.0237)
MAE: 0.3241 (std: 0.0133)
Fit time: 0.03
Test time: 0.00

###SVDpp
RMSE: 0.5238 (std: 0.0168)
MAE: 0.2969 (std: 0.0068)
Fit time: 4.54
Test time: 0.36

###NMF
RMSE: 0.5346 (std: 0.0319)
MAE: 0.2656 (std: 0.0209)
Fit time: 0.07
Test time: 0.00

З отриманих результатів видно що кожна модель має дуже наближені значення RMSE та також схожі значення MAE. Значить кожну моделей можливо використовувати. Але SVDpp має більший за всіх час навчання та тестування.

In [12]:
trainset = data.build_full_trainset()
svd.fit(trainset)

In [13]:
trainset = data.build_full_trainset()
svdpp.fit(trainset)

In [14]:
trainset = data.build_full_trainset()
nmf.fit(trainset)

In [15]:
ratings_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/ratings.csv')
movies_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/movies.csv')

In [16]:
users = ratings_data['userId'].unique()
movies = ratings_data['movieId'].unique()

In [17]:
num_users = len(users)
num_movies = len(movies)

In [18]:
ratings_matrix = np.zeros((num_users, num_movies))

In [19]:
user_indices = np.searchsorted(users, ratings_data['userId'])
movie_indices = np.searchsorted(movies, ratings_data['movieId'])

In [20]:
valid_indices = (user_indices < num_users) & (movie_indices < num_movies)
user_indices = user_indices[valid_indices]
movie_indices = movie_indices[valid_indices]
ratings_data = ratings_data[valid_indices]

In [21]:
ratings_matrix[user_indices, movie_indices] = ratings_data['rating']

In [22]:
user_similarity = np.dot(ratings_matrix, ratings_matrix.T)

In [24]:
def predict_rating(user_id, movie_id):
    user_index = np.where(users == user_id)[0][0]
    movie_index = np.where(movies == movie_id)[0][0]

    rated_users = np.nonzero(ratings_matrix[:, movie_index])[0]

    similarities = user_similarity[user_index, rated_users]

    k = 5
    top_similarities = np.argsort(similarities)[-k:]

    weighted_ratings = np.sum(similarities[top_similarities] * ratings_matrix[rated_users[top_similarities], movie_index])
    similarity_sum = np.sum(np.abs(similarities[top_similarities]))

    if similarity_sum > 0:
        predicted_rating = weighted_ratings / similarity_sum
    else:
        predicted_rating = 0.0

    movie_title = movies_data[movies_data['movieId'] == movie_id]['title'].values[0]

    return predicted_rating, movie_title

In [25]:
user_id = 1
movie_id = 665

predicted_rating, movie_title = predict_rating(user_id, movie_id)
print(f"Predicted user rating {user_id} for movie '{movie_title} (ID: {movie_id})': {predicted_rating}")

Прогнозований рейтинг користувача 1 для фільму 'Underground (1995) (ID: 665)': 3.3922651933701657


In [26]:
user_id = 2
movie_id = 2404

predicted_rating, movie_title = predict_rating(user_id, movie_id)
print(f"Predicted user rating {user_id} for movie '{movie_title} (ID: {movie_id})': {predicted_rating}")

Прогнозований рейтинг користувача 2 для фільму 'Rambo III (1988) (ID: 2404)': 5.0


In [27]:
def get_top_movies_for_user(user_id, top_n=10):
    user_index = np.where(users == user_id)[0][0]

    predicted_ratings = np.zeros(num_movies)
    for movie_index in range(num_movies):
        if ratings_matrix[user_index, movie_index] != 0:
            continue

        similarities = user_similarity[user_index, :]
        rated_users = np.nonzero(ratings_matrix[:, movie_index])[0]
        weighted_ratings = np.sum(similarities[rated_users] * ratings_matrix[rated_users, movie_index])
        similarity_sum = np.sum(np.abs(similarities[rated_users]))
        if similarity_sum > 0:
            predicted_ratings[movie_index] = weighted_ratings / similarity_sum

    top_movie_indices = np.argsort(predicted_ratings)[::-1][:top_n]

    top_movie_titles = [movies_data[movies_data['movieId'] == movies[index]]['title'].values[0] for index in top_movie_indices]

    return top_movie_indices, top_movie_titles

In [28]:
user_id = 2
top_movie_indices, top_movie_titles = get_top_movies_for_user(user_id)

print(f"Top 10 movies for user - {user_id}:")
for index, title in zip(top_movie_indices, top_movie_titles):
    print(f"{title} (ID: {movies[index]})")

Топ-10 фільмів для користувача 2:
When We Were Kings (1996) (ID: 1147)
Rambo III (1988) (ID: 2404)
Falling Down (1993) (ID: 3020)
Chopper (2000) (ID: 4251)
Hudsucker Proxy, The (1994) (ID: 471)
Farewell My Concubine (Ba wang bie ji) (1993) (ID: 446)
Onegin (1999) (ID: 3161)
La Cérémonie (1995) (ID: 1406)
Perfect Candidate, A (1996) (ID: 1123)
Glass Menagerie, The (1987) (ID: 4113)
